src/Raven.Server/Documents/Indexes/StaticIndexHelper.cs

```diff
@@ -140,22 +203,39 @@ private static bool IsStaleDueToReferences(Index index, AbstractStaticIndexBase
+                        var lastProcessedReferenceEtag = index._indexStorage.ReferencesForDocuments.ReadLastProcessedReferenceEtag(indexContext.Transaction, collection, referencedCollection);
+
+                        var lastTombstoneEtag = queryContext.Documents.DocumentDatabase.DocumentsStorage.GetLastTombstoneEtag(queryContext.Documents, referencedCollection.Name);
+                        var lastProcessedTombstoneEtag = index._indexStorage.ReferencesForDocuments.ReadLastProcessedReferenceTombstoneEtag(indexContext.Transaction, collection, referencedCollection);
+
+                        *(long*)writePos = lastDocEtag;
+                        writePos += sizeof(long);
+                        *(long*)writePos = lastProcessedReferenceEtag;
+                        writePos += sizeof(long);
+                        *(long*)writePos = lastTombstoneEtag;
+                        writePos += sizeof(long);
+                        *(long*)writePos = lastProcessedTombstoneEtag;
+                    }
+                }
 
-                foreach (var referencedCollection in referencedCollections)
+                if (compiled.CollectionsWithCompareExchangeReferences.Contains(collection))
                 {
-                    var lastDocEtag = documentsContext.DocumentDatabase.DocumentsStorage.GetLastDocumentEtag(documentsContext, referencedCollection.Name);
-                    var lastProcessedReferenceEtag = index._indexStorage.ReadLastProcessedReferenceEtag(indexContext.Transaction, collection, referencedCollection);
+                    var lastCompareExchangeEtag = queryContext.Documents.DocumentDatabase.ServerStore.Cluster.GetLastCompareExchangeIndexForDatabase(queryContext.Server, queryContext.Documents.DocumentDatabase.Name);
+                    var lastProcessedReferenceEtag = index._indexStorage.ReferencesForCompareExchange.ReadLastProcessedReferenceEtag(indexContext.Transaction, collection, referencedCollection: IndexStorage.CompareExchangeReferences.CompareExchange);
 
-                    var lastTombstoneEtag = documentsContext.DocumentDatabase.DocumentsStorage.GetLastTombstoneEtag(documentsContext, referencedCollection.Name);
-                    var lastProcessedTombstoneEtag = index._indexStorage.ReadLastProcessedReferenceTombstoneEtag(indexContext.Transaction, collection, referencedCollection);
+                    var lastTombstoneEtag = queryContext.Documents.DocumentDatabase.ServerStore.Cluster.GetLastCompareExchangeTombstoneIndexForDatabase(queryContext.Server, queryContext.Documents.DocumentDatabase.Name);
+                    var lastProcessedTombstoneEtag = index._indexStorage.ReferencesForCompareExchange.ReadLastProcessedReferenceTombstoneEtag(indexContext.Transaction, collection, referencedCollection: IndexStorage.CompareExchangeReferences.CompareExchange);
 
-                    *(long*)writePos = lastDocEtag;
+                    *(long*)writePos = lastCompareExchangeEtag;
```

Comment:
How do you ensure we have enough space for this?

Reply:
Handled in `CalculateIndexEtag`

```
            if (_handleCompareExchangeReferences != null)
                length += sizeof(long) * 4 * Collections.Count;
```

Actually I can make it smaller. It can be 
```
            if (_handleCompareExchangeReferences != null)
                length += sizeof(long) * 4 * _compiled.CollectionsWithCompareExchangeReferences.Count;
```